go/arrow/util/protobuf_reflect.go

```diff
@@ -0,0 +1,447 @@
+
+func (psr ProtobufStructReflection) getFieldByName(n string) ProtobufFieldReflection {
+	fd := psr.descriptor.Fields().ByTextName(xstrings.ToSnakeCase(n))
+	fv := psr.rValue
+	if fv.IsValid() {
+		if !fv.IsZero() {
+			for fv.Kind() == reflect.Ptr || fv.Kind() == reflect.Interface {
+				fv = fv.Elem()
+			}
+			if fd.ContainingOneof() != nil {
+				n = string(fd.ContainingOneof().Name())
+			}
+			fv = fv.FieldByName(xstrings.ToCamelCase(n))
+			for fv.Kind() == reflect.Ptr {
+				fv = fv.Elem()
+			}
+		}
+	}
+	return ProtobufFieldReflection{
+		fd,
+		psr.message.Get(fd),
+		fv,
+		psr.SchemaOptions,
+	}
+}
+
+type ProtobufFieldReflection struct {
+	descriptor protoreflect.FieldDescriptor
+	prValue    protoreflect.Value
+	rValue     reflect.Value
+	SchemaOptions
+}
+
+func (pfr ProtobufFieldReflection) isStruct() bool {
+	return pfr.descriptor.Kind() == protoreflect.MessageKind && !pfr.descriptor.IsMap() && pfr.rValue.Kind() != reflect.Slice
+}
+
+func (pfr ProtobufFieldReflection) isMap() bool {
+	return pfr.descriptor.Kind() == protoreflect.MessageKind && pfr.descriptor.IsMap()
+}
+
+func (pfr ProtobufFieldReflection) isList() bool {
+	return pfr.descriptor.IsList() && pfr.rValue.Kind() == reflect.Slice
+}
+
+func (pfr ProtobufFieldReflection) getListLength() int {
+	return pfr.prValue.List().Len()
+}
+
+func (pfr ProtobufFieldReflection) getMapLength() int {
+	return pfr.prValue.Map().Len()
+}
+
+func (plr ProtobufListReflection) generateListItems() chan ProtobufFieldReflection {
+	out := make(chan ProtobufFieldReflection)
+
+	go func() {
+		defer close(out)
+		for i := 0; i < plr.prValue.List().Len(); i++ {
+			out <- ProtobufFieldReflection{
+				descriptor:    plr.descriptor,
+				prValue:       plr.prValue.List().Get(i),
+				rValue:        plr.rValue.Index(i),
+				SchemaOptions: plr.SchemaOptions,
+			}
+		}
+	}()
+
+	return out
+}
+
+func (pfr ProtobufFieldReflection) getDataType() arrow.DataType {
+	var dt arrow.DataType
+
+	typeMap := map[protoreflect.Kind]arrow.DataType{
+		//Numeric
+		protoreflect.Int32Kind:    arrow.PrimitiveTypes.Int32,
+		protoreflect.Int64Kind:    arrow.PrimitiveTypes.Int64,
+		protoreflect.Sint32Kind:   arrow.PrimitiveTypes.Int32,
+		protoreflect.Sint64Kind:   arrow.PrimitiveTypes.Int64,
+		protoreflect.Uint32Kind:   arrow.PrimitiveTypes.Uint32,
+		protoreflect.Uint64Kind:   arrow.PrimitiveTypes.Uint64,
+		protoreflect.Fixed32Kind:  arrow.PrimitiveTypes.Uint32,
+		protoreflect.Fixed64Kind:  arrow.PrimitiveTypes.Uint64,
+		protoreflect.Sfixed32Kind: arrow.PrimitiveTypes.Int32,
+		protoreflect.Sfixed64Kind: arrow.PrimitiveTypes.Int64,
+		protoreflect.FloatKind:    arrow.PrimitiveTypes.Float32,
+		protoreflect.DoubleKind:   arrow.PrimitiveTypes.Float64,
+		//Binary
+		protoreflect.StringKind: arrow.BinaryTypes.String,
+		protoreflect.BytesKind:  arrow.BinaryTypes.Binary,
+		//Fixed Width
+		protoreflect.BoolKind: arrow.FixedWidthTypes.Boolean,
+		// Enum
+		protoreflect.EnumKind: arrow.PrimitiveTypes.Int32,
```

Comment:
should an enum be a dictionary type using the strings as keys?

Reply:
I wasn't familiar with the arrow dictionary type, I'll take a look at it.

Reply:
In order to get the enum string representation I think you need to do a lookup, e.g. 
```
AllTheTypes_ExampleEnum_name[AllTheTypes_OPTION_0]
```
I am not sure how to do that with reflection (or if it is possible).
If it is trivial for you, then please let me know how, otherwise I am happy to only have access to the enum int value.

Reply:
I have had a go at this, but I am writing the string "enum"s as values of the dictionary. I'm not 100% sure this is the functionality you want as we are not preserving the int "enum" anymore.

Reply:
and tbh, i'm not sure why a dict is better than a map here.